### DISCUSSION TITLE
leak report

### DISCUSSION THREAD

**Discussion by tbrekalo** [wjakob/nanobind#13] [Q&A] [answered]
How do I interpret the following?

```
nanobind: leaked 1 types!
 - leaked type "Pile"
nanobind: leaked 5 functions!
 - leaked function "<anonymous>"
 - leaked function "<anonymous>"
 - leaked function "<anonymous>"
 - leaked function "<anonymous>"
 - leaked function "__init__"
nanobind: this is likely caused by a reference counting issue in the binding code.
```

Is it a bug on my part for getting something wrong in binding code or is it an issue with nanobind? 
Side note: code causing this issue deals with shared pointers for another type that is not brought up by the error message. 
Type 'Pile' is regular value type, I have a std::vector of them. So I am not sure how reference counting is being mentioned here.

Thanks for your help in advance.

**Comment by wjakob** [comment]
Yes, this most likely means that there is a bug in your program. I recommend minimizing the bindings as much as possible to produce a tiny reproducer that could be posted here. Probably you will realize what the problem is along the way.

**Reply by tbrekalo** [reply]
Hi Wenzel, thanks for the reply. :) The type leak issue seems to be related to python's typing hints.
```python
tp_handle = camelpy.create_thread_pool(1)
piles = camelpy.deserialize_piles(tp_handle, './test_dump')
# piles: List[camelpy.Pile] = camelpy.deserialize_piles(tp_handle, './test_dump')
```

Commeting out the second line and using the last one seems to trigger the error. 
Is nanobind expected to be compatible with python type hints?

**Reply by wjakob** [reply]
Uh oh, that's not good. If it really boils down to that line, then I'd say that Python's typing hints are responsible. nanobind's shutdown routine (which emits these warnings) gets called very late during interpreter shutdown.  At this point, Python should have garbage collected all types defined by your binding code, and a warning indicates that this hasn't happened. So something is still holding a reference.

However, I cannot easily reproduce your issue with the project-provided test suite. For example, I tried adding

```
from typing import List
s1 : List[t.Struct] = None
```

at the end of `tests/test_classes.py`, and this doesn't trigger the reference counting warning.

**Reply by wjakob** [reply]
So this is a request: can you modify the test suite with a minimal addition to produce such warning? It would also be helpful to know your exact OS and Python version.

**Reply by tbrekalo** [reply]
Do you want me to post a .zip file with modified [nanobind_example](https://github.com/wjakob/nanobind_example) or make a fork and link it here? :) 

I am testing the code on a remote machine; Ubuntu 20.04.3 LTS, Python 3.9.9, GCC 9.4.0
Will try replicating bug locally on macos 12.2.1; Python 3.9.10, Apple Clang 13.0

**Reply by wjakob** [reply]
If it's a modification of the default `nanobind_example`, you could simply post the output of `git diff` here.

**Reply by tbrekalo** [reply]
Unfortunatelly I am unable to reproduce bug in the same environment on the small scale. Having hard time figuring out how the interaction between my code and nanobind influnces python type hints. I have other more urgent things to resolve but If I manage to track down the issues in the near future; I'll post it here. 

Thanks for you support.

**Reply by tbrekalo** [reply]
Managed to construct a minimum program which triggers the error with slight modifications to [nanobind_example](https://github.com/wjakob/nanobind_example).

Here's the git diff:
```bash
index 95ecfec..7d3562a 100644
--- a/src/nanobind_example/__init__.py
+++ b/src/nanobind_example/__init__.py
@@ -1 +1 @@
-from .nanobind_example_ext import add
+from .nanobind_example_ext import add, Blob
diff --git a/src/nanobind_example_ext.cpp b/src/nanobind_example_ext.cpp
index 602ee79..c3d65f7 100644
--- a/src/nanobind_example_ext.cpp
+++ b/src/nanobind_example_ext.cpp
@@ -4,6 +4,13 @@ namespace nb = nanobind;

 using namespace nb::literals;

+struct Blob {
+  Blob(int a) : a(a) {}
+  int a;
+};
+
 NB_MODULE(nanobind_example_ext, m) {
-    m.def("add", [](int a, int b) { return a + b; }, "a"_a, "b"_a);
+  nb::class_<Blob>(m, "Blob").def(nb::init<int>());
+  m.def(
+      "add", [](int a, int b) { return a + b; }, "a"_a, "b"_a);
 }
diff --git a/tests/test_basic.py b/tests/test_basic.py
index 9c670fc..285c927 100644
--- a/tests/test_basic.py
+++ b/tests/test_basic.py
@@ -1,4 +1,9 @@
 import nanobind_example as m
+import seaborn as sns

-def test_add():
+from typing import List
+
+def test_add():
     assert m.add(1, 2) == 3
+
+blobs: List[m.Blob] = []
```

If I exclude the: `import seaborn as sns` from the test file, nanobind doesn't raise the issue for the `Blob` type.

**Reply by jbms** [reply]
In general, Python does not guarantee that all references are cleaned up when the interpreter is shut down.  For example, in the case of multiple threads, I believe leaks are practically guaranteed.

For details you can review the CPython source code:
https://github.com/python/cpython/blob/75280944e5ca957eec7f814b9d0608fc84fc5811/Python/pylifecycle.c#L1749

**Reply by tbrekalo** [reply]
Then should an issue be created from this discussion?

**Reply by wjakob** [reply]
Yes, that would be great. I would generally say that these warning messages are far too useful to simply turn them off. (I already found a ton of bugs with them, both in my own projects, and in nanobind itself). However, perhaps they can be "restrained" somehow so that they are aware of potential leaks involving type annotations.

**Reply by tbrekalo** [reply]
Opened an issues from this discussion. Would love to help resolving this issue but I am afraid my knoweledge of python inner workings is too shallow. Maybe having an option to disable warnings for objects whoes ownership is not shared between the binding code and python interpreter?

**Comment by wjakob** [comment, marked as answer]
For anyone running into this in the future: the problem arises due to a reference leak in `typing` (not exactly this issue, but this is what pointed me into the right direction: https://bugs.python.org/issue28649).

To avoid this issue, a set of LRU caches related to type information must be cleared before interpreter shutdown, which can be accomplished by running the following anywhere (just once is enough).
```
# Uncomment to stop leak warnings
import atexit

def cleanup():
    import typing
    for cleanup in typing._cleanups:
        cleanup()

atexit.register(cleanup)
```
 In any case, this is not a _nanobind_ issue. It's just that _nanobind_ complains very loudly about leaks (a good thing!) which revealed this issue.
